### DISCUSSION TITLE
What is the process to load schemas with multi level references.

### DISCUSSION THREAD

**Discussion by lekhamaru** [Apicurio/apicurio-registry#5493] [Q&A]
Hello Experts

I have a json schema which has a reference hierarchy as follows

Characteristic --> CharacteristicRelationship --> Extensible

I'm using Maven plugin to upload artifacts with references. I uploaded Extensible schema first. Then CharacteristicRelationship and then Characteristic schema. But references to the 2nd level are lost when I upload the Characteristic artifact. I can no longer see the reference between the 2nd and 3rd level.  Please can you hep with the correct way to upload.

**Comment by Ganesh-403** [comment]
From what I can tell, **the reference relationships should be preserved**. A parent artifact referencing `CharacteristicRelationship`, which in turn references `Extensible`, should result in each artifact retaining its own direct references.

For example:

```
Characteristic
    └── CharacteristicRelationship
            └── Extensible
```

should produce:

- `Characteristic` → `CharacteristicRelationship`
- `CharacteristicRelationship` → `Extensible`

The Registry stores references **per artifact**, so registering `Characteristic` should not overwrite or remove the references that already belong to `CharacteristicRelationship`. The Maven plugin documentation also describes references as artifact-level metadata, and nested/multi-level references are a supported concept for JSON Schema. :contentReference[oaicite:0]{index=0}

A couple of things to verify:

1. **How are you registering the references?**
   - Using explicit `<references>` entries?
   - Or using `<autoRefs>true</autoRefs>`?

   `autoRefs` is currently a **Technology Preview** feature, so if you're using that, it's possible you've encountered a limitation or bug in the automatic detection logic. :contentReference[oaicite:1]{index=1}

2. **Are you updating an existing artifact?**

   If you re-register `CharacteristicRelationship` without its original `references` metadata, the Registry will update that artifact version with the references supplied in the new registration. Make sure every registration of an artifact includes all of its direct references.

3. **Verify the intermediate artifact after each step.**

   After registering `CharacteristicRelationship`, check:

   ```text
   GET /apis/registry/v3/groups/{groupId}/artifacts/CharacteristicRelationship/versions/{version}/references
   ```

   It should still list the reference to `Extensible`.

   Then, after registering `Characteristic`, check the same endpoint again. If the reference has disappeared, that would suggest a bug, since registering one artifact should not modify another artifact's reference metadata. :contentReference[oaicite:2]{index=2}

I couldn't find any documentation stating that multi-level references are flattened or that second-level references are intentionally discarded. Based on the documented behavior, I would expect each artifact to maintain its own direct references independently.

Could you share:

- whether you're using manual `<references>` or `autoRefs`,
- the relevant portion of your `pom.xml`,
- and the output of the `/references` endpoint for `CharacteristicRelationship` **before and after** registering `Characteristic`?

That would help determine whether this is a configuration issue or a bug in the Maven plugin or Registry.

If this solves your problem, feel free to mark it as the accepted answer so others can find it easily.
@lekhamaru
